feat: support for multiple entry points / multi-project support

michaelbe812 opened the pull request:
PR for implementing support for multi-project workspaces.


# Feature
Running `sheriff list` in a multi-workspace project like the test-project angular-v-multi will now output:

```sh
This project contains 10 modules:

. (root)
└── projects
  ├── app-i
    └── src
      └── app
        └── non-compliant
          ├── data-access (type:data-access)
          ├── feat (type:feat)
          ├── types (type:types)
          ├── ui (type:ui)
          └── util (type:util)
  └── app-ii
    └── src
      └── app
        └── non-compliant
          ├── data-access (type:data-access)
          ├── feat (type:feat)
          ├── types (type:types)
          ├── ui (type:ui)
          └── util (type:util)
```
For single project workspaces nothing changed.

`sheriff verify` will output in case of violotions for multi-project workspaces:

```sh
Verification Report

Project: app-i

Issues found:
  Total Invalid Files: 4
  Total Encapsulation Violations: 1
  Total Dependency Rule Violations: 5
----------------------------------

|-- projects/app-i/src/app/app.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag root to tags type:feat
|   |   |-- from tag root to tags type:ui
|-- projects/app-i/src/app/non-compliant/feat/feat-violation.component.ts
|   |-- Encapsulation Violations
|   |   |-- ../util/internal/internal-util
|-- projects/app-i/src/app/non-compliant/ui/ui-violation.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag type:ui to tags type:data-access
|   |   |-- from tag type:ui to tags type:feat
|-- projects/app-i/src/app/non-compliant/types/violation.types.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag type:types to tags type:util

Project: app-ii

Issues found:
  Total Invalid Files: 4
  Total Encapsulation Violations: 1
  Total Dependency Rule Violations: 5
----------------------------------

|-- projects/app-ii/src/app/app.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag root to tags type:feat
|   |   |-- from tag root to tags type:ui
|-- projects/app-ii/src/app/non-compliant/feat/feat-violation.component.ts
|   |-- Encapsulation Violations
|   |   |-- ../util/internal/internal-util
|-- projects/app-ii/src/app/non-compliant/ui/ui-violation.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag type:ui to tags type:data-access
|   |   |-- from tag type:ui to tags type:feat
|-- projects/app-ii/src/app/non-compliant/types/violation.types.ts
|   |-- Dependency Rule Violations
|   |   |-- from tag type:types to tags type:util
```

If no violations are found:
```sh 
Verification Report

Project: app-i


No issues found for this project. Well done!

Project: app-ii


No issues found for this project. Well done!

All projects validated successfully!

```

For single project workspaces nothing changed

# Hint for Reviewer to Test
Kindly run the sheriff tasks in the new `test-project` `angular-v-multi`

michaelbe812 commented on packages/core/src/lib/cli/internal/get-entry-from-cli-or-config.ts:
```diff
@@ -26,3 +41,34 @@ export function getEntryFromCliOrConfig(entryFile = '', runInit = true): Project
+    return runInit
+      ? [
+          {
+            projectName: DEFAULT_PROJECT_NAME,
```
hint for reviewer
For single project workspaces I am assigning here a constant for the project name so that I can later in e.g. verify easily know if where are inside a single or multi project workspace

michaelbe812 commented on packages/core/src/lib/config/user-sheriff-config.ts:
```diff
@@ -220,20 +220,20 @@ export interface UserSheriffConfig {
+   * The file or files that the CLI should use by default.
    */
-  entryFile?: string;
+  entryFile?: string | Record<string, string>;
```
**Question**

I did not wanted to create a separate property `entryFiles`. Imho a union is more convenient here. 

We might want to update the name. 

I was also thinking about to make `entryFile` only accept a `Record<string,string>`. This would mean for single project workspaces a tiny overhead but we could simplify a bit the internal implementation of sheriff.

What do you think?

michaelbe812 commented:
**Question**

I am not sure if I added correctly the new test-project. Please double-check here :)

rainerhahnekamp commented:
Thanks Michael, I will start the review ASAP (but could take a while)

michaelbe812 commented:
> Thanks Michael, I will start the review ASAP (but could take a while)

Looking forward to:) 
Just take your time 👍🏻

rainerhahnekamp commented:
Sorry, 2 weeks have passed, and I still haven't found the time for the review. What came to mind, though, were some suggestions for the API.

- I see you have a union type with Records. What if we differentiate between `entryFile` and `entryPoints`? `entryPoints` could be more than one element, and their keys aren't the files but just a name. That's also why I wouldn't call them `entryFiles`.
- We should also foresee an optiontor to define a default `entryPoint`. With `entryFile` that is unnecessary but with `entryPoints` it might be a suitable, optional property.

michaelbe812 commented:
No worries Rainer, just take you time. I might also have some longer response times in the next weeks because of other priorities and duties :-) 

Inline your suggestions, makes absolutely sense. 

To summit up: 
I will revert back to `entryFile?:string` 

And add another option `entryPoints?: Record<string,string>`. 

I guess it only makes sense to have either `entryFile` or `entryPoints` given, right? So I would also validate that. 

Maybe we can also come up with a more self-explanatory/ streamlined API for the different workspace types (I have no idea yet). 

What's the idea behind the `defaultEntryPoint` option?

michaelbe812 commented:
> No worries Rainer, just take you time. I might also have some longer response times in the next weeks because of other priorities and duties :-)
> 
> Inline your suggestions, makes absolutely sense.
> 
> To summit up: I will revert back to `entryFile?:string`
> 
> And add another option `entryPoints?: Record<string,string>`.
> 
> I guess it only makes sense to have either `entryFile` or `entryPoints` given, right? So I would also validate that.
> 
> Maybe we can also come up with a more self-explanatory/ streamlined API for the different workspace types (I have no idea yet).
> 
> What's the idea behind the `defaultEntryPoint` option?

Just as info

I already incorporated your feedback regardign `entryFile`. I reverted the change and introduced `entryPoints` as separate config option. 

I did also add a few test cases for multi-project setups

michaelbe812 commented:
> @michaelbe812 thanks for letting me know. I will have more time in the upcoming two weeks. Maybe I manage to do the review over the weekend, but no gurantee

Don't feel pressured 😇 by accident I just had a bit of time. 

Looking forward to your feedback:)

rainerhahnekamp reviewed (changes requested):
Hey @michaelbe812,

Thanks again for this one — it's going to be an important feature and will
really improve the combination of Nx and Sheriff.

I left a few comments regarding implementation details, but I would also like
to discuss how we handle `entryPoints` internally.

Currently, a large part of Sheriff processes only a single `entryPoint`,
meaning we don't need to check whether other `entryPoints` exist.

My idea would be that `entryPoints` should not be used after the configuration
has been parsed. During parsing, we should determine whether the `entryFile`
comes directly from `entryFile` or is derived from `entryPoints`.

To support this, we would also need to adjust the `Configuration`.  
I suggest removing `entryFile` and introducing something like `finalEntryFile`
(or a better name if you have one).

The renaming should make it clear that this is not simply the `entryFile` from
the configuration, but a resolved value — either specified directly or derived
from `entryPoints`. A small comment in the code would likely be helpful to make
this clear.

`parseConfig` would be responsible for resolving this field.  
As a result, all the CLI changes you introduced would not be necessary anymore
— the only change would be to use `finalEntryFile` (or whatever name we choose)
instead of `entryFile`.

I would still keep `entryPoints` in the `Configuration`, because I imagine
that in the future, CLI commands like `verify`, `export`, and `list` could offer
an option to run through all `entryPoints`.  
However, I would treat that as a separate PR.

What do you think about this approach?

---

Regarding the review:  
- I checked only the core changes for now and haven't looked into
`angular-v-multi` yet.  
- I expect there will be more review rounds, as this is a complex topic — but
I'll try to respond quicker next time!

---

**Summary**

- [ ] Decide on the `finalEntryFile` (or alternative name) and implement it.
- [ ] Update documentation accordingly.

michaelbe812 commented:
To keep you updated Rainer: 
I plan to continue work this week

michaelbe812 commented:
> @ michaelbe812
> 
> When I wrote the initial review, I hadn't considered the option of running all entry points by default. But after you showed it to me, I had some time to think it through — and I now believe that running all when no `entryPoint` is specified is actually the best approach. Since `entryPoints` is a new feature, we’re not breaking any existing behavior.
> 
> As for the `--default` flag: is it even necessary anymore? If the user already has to provide a name, it doesn’t seem to make a big difference whether it's `--default` or just the project name.
> 
> Let me know what you think.

I would probably keep the behavior as it is right now. 

If wished we could for instance in the future add a CLI argument `--entryPoint=app1` where `app1` is a key of `entryPoints`.

michaelbe812 commented:
@rainerhahnekamp 
I think the final review can be done.

Kindly test thorougly on your own the new functionality.

Actually I do not know why the Pipeline is failing. For me it looks like the `run-integration-tests` is running fine. But maybe I am missing something

michaelbe812 commented:
> @michaelbe812, just to be sure. we are here on the same page.
> 
> no `entryFile`, no `entryPoints`:
> 
> ```ts
>  // will fail and demand a file as parameter
> npx sheriff verify // ⛔️
> ```
> 
> `entryFile`, no `entryPoints`:
> 
> ```ts
>  // will work and use entryFile
> npx sheriff verify // ✅
> 
> // will run src/main.ts - even if entryFile is different
> npx sheriff verify src/main.ts // ✅
> ```
> 
> no `entryFile`, `entryPoints`
> 
> ```ts
> // will run on entryPoints
> npx sheriff verify // ✅
> 
> // run entryPoint with that that name -> if holidays exist in entryPoints
> npx sheriff verify holidays // ✅
> 
> // fallback to file if entryPoint  for apps/hotels/src/main.ts does not exist - I think that this is a new requirement
> npx sheriff verify apps/hotels/src/main.ts // ✅
> ```
> 
> Please confirm or correct me. At the moment, I don't see a need for `defaultEntryPoint`.

Regarding the need of `defaultEntryPoint`: 
I do totally agree here and without an additional CLI argument I also do not see the possibility to implement it in a reasonable way.

Regaring the examples you provided:
It should work exactly like this (however I have to double-check if it is implemented like that as I do not remember it by heart :))

michaelbe812 commented:
@rainerhahnekamp 
I just had the time to check all the cases [you mentioned](https://github.com/softarc-consulting/sheriff/pull/199#issuecomment-3000041570).

It all works as expected. 

The only option which is not available atm is `npx sheriff verify holidays` so passing a `entryPoint`.

However I did already see this requirement and opened an [issue yesterday](https://github.com/softarc-consulting/sheriff/issues/214)

I would like to do this after this PR is merged if it is okay for you? 

However I do not see why the build pipeline is failing, could you assist here?

I think we can soon get this PR over the line 🚀

michaelbe812 commented on packages/core/src/lib/cli/list.ts:
```diff
@@ -1,31 +1,35 @@
 import { cli } from './cli';
 import { logInfoForMissingSheriffConfig } from './internal/log-info-for-missing-sheriff-config';
 
 export function list(args: string[]) {
```
done

rainerhahnekamp replied:
please also create a test, which covers multi projects for list

michaelbe812 commented on packages/core/src/lib/cli/export-data.ts:
```diff
@@ -1,12 +1,16 @@
 
-  const data = getProjectData(entryFile, fs.cwd(), { includeExternalLibraries: true });
-  cli.log(JSON.stringify(data, null, '  '));
+  for (const entry of projectEntries) {
```
@rainerhahnekamp 
As far as I checked everything correctly we will need to pass either `Entry` to `getProjectData` instead of the entryPath or add another argument `projectName`.

Nevertheless this would introduce a breaking change because `getProjectData` is exported from the main `index.ts`.

How should be handle this?

rainerhahnekamp replied:
I am not sure if `getProjectData` should be aware of entryPoints vs entryFile. In the end this is really an API and not a CLI command.

michaelbe812 replied:
Then I don't know what to do here

rainerhahnekamp commented on packages/core/src/lib/cli/internal/get-entries-from-cli-or-config.ts:
```diff
@@ -0,0 +1,109 @@
+  if (fs.exists(potentialConfigFile)) {
+    const sheriffConfig = parseConfig(potentialConfigFile);
+
+    if (sheriffConfig.entryFile && !isEmptyRecord(sheriffConfig.entryPoints)) {
```
I don't know. Sometimes doing things like that in the type system can make the code more complicated than it has to be. On the other side, it would be a direct benefit for users.

I leave it up to you. The most important thing is that we have the runtime check.

michaelbe812 replied:
I am with you. 
I will leave it as it is for now.

As it is just some internal type we can do a refactoring anytime we feel it should be the other way around.

michaelbe812 replied:
done

rainerhahnekamp commented on packages/core/src/lib/cli/internal/get-entries-from-cli-or-config.ts:
```diff
@@ -0,0 +1,109 @@
+
+export const DEFAULT_PROJECT_NAME = 'default';
+
+export function getEntriesFromCliOrConfig<R extends boolean = true>(
```
in my experience, if you just rename the file commit, and then change the file and do a second commit, it should work. If you do both in one git commit, it could cause issues.

rainerhahnekamp replied:
Hey, so I think that there is still an issue, where we would loose the history. Is it possible, that I give it a try? Meaning, I would push to your branch?

<img width="327" alt="image" src="https://github.com/user-attachments/assets/c8343875-f5bc-48fd-878c-aa0841de6339" />

rainerhahnekamp commented on packages/core/src/lib/test/is-empty-record.spec.ts:
```diff
@@ -0,0 +1,41 @@
+  it('should return false if value is undefined', () => {
+    expect(isEmptyRecord(undefined)).toBe(false);
+  });
+
```
if you use new line to create groups of tests, you can also use nested describes. In that way you can give your group also a name.

michaelbe812 replied:
That was a co-incidence sorry. I did not intentionally wanted to introduce grouping

michaelbe812 commented:
@rainerhahnekamp 
I did address all your review points (I think at least).

Still the integration-test for `angular-i` is failing.
I dont know why and it would be great if you could assist here.

I did [left a comment here](https://github.com/notifications#discussion_r2173509821) regarding the `export-data`

and also [here](https://github.com/notifications#discussion_r2173852623) because of some typescript issues when making `isEmptyRecord` more generic.

From what I can see these are the last points which are open.

michaelbe812 commented on packages/core/src/lib/cli/internal/is-empty-record.ts:
```diff
@@ -0,0 +1,3 @@
+export function isEmptyRecord<T extends object>(arg: T): boolean{
```
Done

rainerhahnekamp reviewed (commented):
Thanks so much for your patience throughout this process — I really appreciate it. I know we had multiple rounds of review, and it took me way too long to get back to you each time. Thanks again for bearing with me.

I’ve now gone through the code — only a few very minor things, so I believe this will be the final review on my end. Really solid work overall!

Just two final points:
	1.	Integration tests: I haven’t looked at them yet, but I don’t think that should block the PR.
	2.	Git rename issue: About the file rename (get-entry-from-cli-or-config → get-entries-from-cli-or-config) — I tried multiple approaches (and spent quite a bit of time on it), but none worked reliably.

To avoid further issues, I’d suggest splitting this into two merges:
	•	First, a separate PR that only renames the file.
	•	Then, rebase this branch on top of that so we can merge it cleanly.

Let me know what you think — and again, thanks a lot for your great work and your patience!

rainerhahnekamp commented on packages/core/src/lib/cli/internal/get-entries-from-cli-or-config.ts:
```diff
@@ -0,0 +1,111 @@
+): Array<Entry<ProjectInfo>>;
+export function getEntriesFromCliOrConfig(
+  entryFileOrEntryPoints?: string,
+  runInit?: true,
```
My fault again, didn't see that the first parameter was also optional. Then no further action required.

michaelbe812 commented on docs/docs/cli.md:
```diff
@@ -10,20 +10,29 @@ The core package (@softarc/sheriff-core) comes with a CLI to initialize the conf
 
 Run `npx sheriff init` to create a `sheriff.config.ts`. Its configuration runs with [automatic tagging](./dependency-rules#automatic-tagging), meaning no dependency rules are in place, and it only checks for the module boundaries.
 
-## `verify [main.ts]`
```
done
